### DISCUSSION TITLE
How to utilize pre-release version?

### DISCUSSION THREAD

**Discussion by btylerburton** [open-sdg/open-sdg#2067] [Q&A] [answered]
Hello @brockfanning, our build pipeline is ending support for Ruby 2.x. Is there a way to incorporate the pre-release version while [we await the 2.3.0 release](https://github.com/open-sdg/open-sdg/discussions/2064)

**Comment by brockfanning** [comment, marked as answer]
@btylerburton We do ongoing development in branches in each of the repositories according to the next release plus `-dev`. Since the next release will be 2.3.0, the branches right now are 2.3.0-dev. So, to use the code we have ready for 2.3.0, you would need to adjust a few files to point to these branches.

## Data repository

In the data repository, you would need to adjust two places. First, in `scripts/requirements.txt`, instead of...

```
git+https://github.com/open-sdg/sdg-build@2.2.0
```

...you would change it to:

```
git+https://github.com/open-sdg/sdg-build@2.3.0-dev
```

Next, in `config_data.yml`, instead of...

```
  - class: TranslationInputSdgTranslations
    source: https://github.com/open-sdg/sdg-translations.git
    tag: 2.2.0
```

...you would change it to:

```
  - class: TranslationInputSdgTranslations
    source: https://github.com/open-sdg/sdg-translations.git
    tag: 2.3.0-dev
```

## Site repository

In the site repo you'll need to change two places as well. First, in `Gemfile`, instead of...

```
gem "jekyll-open-sdg-plugins", "2.2.0"
```

...you would change it to:

```
gem 'jekyll-open-sdg-plugins', git: 'https://github.com/open-sdg/jekyll-open-sdg-plugins.git', branch: '2.3.0-dev'
```

(the syntax there is a bit different because we're pointing to a branch on Github instead of a Rubygems.org package)

Next, in `_config.yml`, instead of...

```
remote_theme: open-sdg/open-sdg@2.2.0
```

...you would change it to:

```
remote_theme: open-sdg/open-sdg@2.3.0-dev
```

## Ruby 3

The above is the normal procedure for using the pre-release version. There may be other changes needed in certain situations though. For example, to start using Ruby 3, you'll also need to make some other changes in your site repo. Check the changes in this open pull-request to see the other stuff needed for Ruby 3: https://github.com/open-sdg/open-sdg-site-starter/pull/98/files

**Reply by btylerburton** [reply]
Thank you, @brockfanning. We should be picking this work up in the next couple weeks. Will post back here if we spot any integration issues.

**Reply by btylerburton** [reply]
Hello @brockfanning after following the above, we're seeing an issue with kramdown parser in the build.

```
2023-11-08 20:52:29 INFO [build-jekyll] /usr/local/rvm/gems/ruby-3.1.3/gems/kramdown-1.17.0/lib/kramdown/parser/html.rb:10:in `require': cannot load such file -- rexml/parsers/baseparser (LoadError)

2023-11-08 20:52:29 INFO [build-jekyll] from /usr/local/rvm/gems/ruby-3.1.3/gems/kramdown-1.17.0/lib/kramdown/parser/html.rb:10:in `<top (required)>'

2023-11-08 20:52:29 INFO [build-jekyll] from /usr/local/rvm/gems/ruby-3.1.3/gems/kramdown-1.17.0/lib/kramdown/parser/kramdown/html.rb:10:in `require'

2023-11-08 20:52:29 INFO [build-jekyll] from /usr/local/rvm/gems/ruby-3.1.3/gems/kramdown-1.17.0/lib/kramdown/parser/kramdown/html.rb:10:in `<top (required)>'

2023-11-08 20:52:30 INFO [build-jekyll] Now using node v18.18.0 (npm v9.8.1)

2023-11-08 20:52:30 INFO [build-jekyll] /usr/local/rvm/gems/ruby-3.1.3/gems/kramdown-1.17.0/lib/kramdown/parser/html.rb:10:in `require': cannot load such file -- rexml/parsers/baseparser (LoadError)

2023-11-08 20:52:30 INFO [build-jekyll] from /usr/local/rvm/gems/ruby-3.1.3/gems/kramdown-1.17.0/lib/kramdown/parser/html.rb:10:in `<top (required)>'

2023-11-08 20:52:30 ERROR [main] There was a problem running Jekyll, see the above logs for details.

```

We've added Webrick as a dependency, as specified here: https://github.com/jekyll/jekyll/issues/8523, but the build is still failing. 

Is this something you've encountered?

**Reply by christoph-LSN** [reply]
Hi @btylerburton could you post site and data repo links, perhaps I can take a look.....

**Reply by brockfanning** [reply]
@btylerburton Did you also add `kramdown-parser-gfm` as a dependency? See the link I posted above as an example of the changes needed in the Gemfile.

**Reply by btylerburton** [reply]
I missed that one @brockfanning. Thanks for the tip! 
That got us a bit farther... Now we're seeing build issues during the generating phase... 

```
jekyll 3.9.3 | Error:  undefined method `end_with?' for nil:NilClass
/Users/brendantburton/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/jekyll-open-sdg-plugins-cac676f67ec9/lib/jekyll-open-sdg-plugins/routes.rb:16:in `generate': undefined method `end_with?' for nil:NilClass (NoMethodError)

      unless baseurl.end_with?('/')
```

Using Ruby 3.2.2

**Reply by btylerburton** [reply]
@christoph-LSN here's those links:

site: https://github.com/GSA/sdg-indicators-usa/tree/integrate-pre-release-version_ruby3
staging data: https://github.com/GSA/sdg-data-usa-staging

**Reply by btylerburton** [reply]
Hold the phone, another team member got it working. We just needed to add `baseurl: ""` to our config.yaml

**Reply by brockfanning** [reply]
@btylerburton I also noticed that issue. I've just pushed an update to jekyll-open-sdg-plugins, so I think you should be able to remove the `baseurl: ""` now.

**Reply by btylerburton** [reply]
Nicely done, Brock! 🥳 I believe that should get us to a new 3.x build, but will circle back once we get all things in place.

**Reply by btylerburton** [reply]
We successfully upgraded our staging repo here: https://github.com/GSA/sdg-indicators-usa/pull/1056/files. 

Had to add `observation_attibutes` to get the charts to display FYI. Otherwise it was passing a `null` into the map config.

```
observation_attributes:
  - field: COMMENT_OBS
    label: ''
```

<img width="1330" alt="Screenshot 2023-11-09 at 11 16 02 AM" src="https://github.com/open-sdg/open-sdg/assets/104455281/8e7503a0-54a9-4169-83d5-c0299ec9a314">
